### ISSUE TITLE
Computing running losses in diffmpm

### ISSUE THREAD

**Issue by aespielberg** [yuanming-hu/taichi#264]
I'm trying to add running losses in the diffmpm code; however, I must be misunderstanding something, because I'm getting errors.  In particular, I want my loss now to be a final position plus the integrated position.  Yes, this is redundant, as the two values should be the same up to integration error, but this is purely for the sake of proving the concept.

I add new tensors for v_avg at the top of the file, as well as running_loss and total_loss.  total_loss should be (final) loss + running_loss:

`v_avg = vec() #mean velocity of the robot at each time s`

```
running_loss = scalar()
total_loss = scalar()
```
I want to compute the running loss and v_avg at each time step, so I make `steps` copies in place:

```
ti.root.dense(ti.i, steps).place(running_loss)
ti.root.dense(ti.i, steps).place(v_avg)
```

Next, a couple of helper functions, for computing the running_loss, modeled after those for loss:

```
@ti.kernel
def compute_v_avg(s: ti.i32):
  contrib = 1.0 / n_particles
  for i in range(n_particles):
    v_avg[s].atomic_add(contrib * v[s, i])
```

```
@ti.kernel
def compute_running_loss(s: ti.i32):
  velocity = v_avg[s, 0]
  running_loss[s] = -velocity * dt
```
and one for computing total loss:

```
@ti.kernel
def compute_total_loss():
  total_loss[None] = 0.0
  for i in range(steps):
    total_loss[None].atomic_add(running_loss[s])
  total_loss[None] += loss[None]
```

Finally, we modify forward simulation in order to capture the running loss:

```
def forward(total_steps=steps):
  # simulation
  for s in range(total_steps - 1):
    advance(s)
    v_avg[s] = zero_vec_list()
    compute_v_avg(s)
    compute_running_loss(s)
 
  x_avg[None] = zero_vec_list()
  compute_x_avg()
  compute_loss()
  compute_total_loss()
```

where I've defined the helper function, zerfo_vec_list:

```
def zero_vec_list(dim=dim):
  return [0. for _ in range(dim)]
```

Unfortunately, this is triggering gdb errors early on, in compute_v_avg.  What am I doing wrong?  I suspect I am doing something incorrect with how I am instantiating or accessing these tensors, but I don't understand the errors :(

**Comment by yuanming-hu** [comment]
Sorry about that. I'm not sure what's wrong exactly in your script. Could you share with me the whole script?

**Comment by aespielberg** [comment]
I will share with you shortly via e-mail so as not to clutter up github
with my gross looking code :)
